core/divelist.c

```diff
@@ -166,21 +170,39 @@ static int calculate_otu(const struct dive *dive)
 
 	for (i = 1; i < dc->samples; i++) {
 		int t;
-		int po2;
+		int po2i, po2f;
 		struct sample *sample = dc->sample + i;
 		struct sample *psample = sample - 1;
-		t = sample->time.seconds - psample->time.seconds;
+		t = (sample->time.seconds - psample->time.seconds);
 		if (sample->setpoint.mbar) {
-			po2 = sample->setpoint.mbar;
+			po2i = psample->setpoint.mbar;
+			po2f = sample->setpoint.mbar;
 		} else {
 			int o2 = active_o2(dive, dc, psample->time);
-			po2 = lrint(o2 * depth_to_atm(sample->depth.mm, dive));
+			po2i = lrint(o2 * depth_to_atm(psample->depth.mm, dive));	// (initial) po2 at start of segment
+			po2f = lrint(o2 * depth_to_atm(sample->depth.mm, dive));	// (final) po2 at end of segment
+		}
+		if ((po2i > 500) || (po2f > 500)) {			// 1) If PO2 in segment is above 500 mbar
+			if (psample->depth.mm == sample->depth.mm) {	// ... and this is a constant depth segment
+				otu += t * pow(500.0 / ((double)po2i - 500.0), -0.833333333) / 60.0; // otu if time is minutes
+			} else {					// 2) For dive segments that ascend or descend:
+				if (po2i <= 500) {				// For descent segment with po2i <= 500 mbar ..
+					t = t * (po2f - 500) / (po2f - po2i);	// .. only consider part with PO2 > 500 mbar
+					po2i = 501;
+				} else {
+					if (po2f <= 500){ 
+						t = t * (po2i - 500) / (po2i - po2f);	// For ascent segment with po2f <= 500 mbar ..
+						po2f = 501;				// .. only consider part with PO2 > 500 mbar
+					}
+				}
+				otu += 50.0 * t * (pow(((double)po2f - 500.0) / 500.0, 1.833333333) - pow(((double)po2i - 500.0) / 500.0, 1.833333333))
+					 / (11.0 * (double)(po2f - po2i));	// (if po2 is bar and time is minutes, then (3/11) * (1000/60) = 50/11)
```

Comment:
I have NOT read the original article that you base this on, but the logic here seems quite strange and confusing to me... setting po2i/po2f to 501 and... I simply don't understand what this is trying to achieve.
I'm hoping that @atdotde will review / comment.

Reply:
I believe this is necessary. Oxygen toxicity only becomes an issue for PO2 > 500 mbar. The mathematical (differential calculus) lower limit for oxygen toxicity calculations is 500 mbar. So, for a dive from surface to 30m depth, the part of the descent more shallow than PO2=500 mbar needs to be excluded from the calculation, assuming a constant rate of ascent/descent. For dives (or more especially dive plans) with longish segments this condition is therefore important. I will keep this in mind while I wait for more comments. Thanks for your attention to this.

Reply:
The division by zero is of course bogus: In line 202 you can simply invert the fraction and change the sign of the exponent and it will go away. In line 205, it is cancelled by the nominator that goes like (po2f - po2i)^11/6 in the dangerous region. Let me try to massage these formulas a bit so the don't show these problems. But not now (have to run). In any case, I am looking into the references of the Baker article to track down the origin of the 5/6 power (that I find quite suspicious given this is supposed to be an empirical formula). Maybe we can get away with something simpler. Should get back to you latest tomorrow.

Reply:
So here is what I came up with

double p = (po2f + po2i)/1000.0 - 1.0;

otu += t/60.0 * ( pow(p, 6.0/5.0) + (po2f - po2i)* (po2f -po2i) / 25000000.0 * pow(p, -4.0/5.0));

This works even with po2f == po2i and should be very close numerically to Baker's exact formula (it is a third order approximation in (po2f-po2i)/(po2f + po2i).

I will write a short post about it in my blog but please test that it works numerically.

Reply:
Argh, I had my maths wrong (6/5 instead of 5/6, which makes me think, didn't the original literature have 1 .2 there?). OK, assuming 5/6 is correct, have a look at https://github.com/atdotde/subsurface/blob/c9d550e36fbbe4a3a10a6add808935c82c92efb6/core/divelist.c#L209 This formula seems to get it (sufficiently) right. Have a look at that branch, that is not for merging but it has all the debugging info in it. If you like it, you could take that formula, get rid of the printf and your old formula (that is now with delta_old), and instead fo the delta do out += immediately. Also in the if's further up, it's now fine to have po2 = 500 since we nowhere divide by this anymore.

Still need to write that blog post...